### DISCUSSION TITLE
Allow to specify requested body/params types in defineEventHandler api routes for a typesafe useFetch

### DISCUSSION THREAD

**Discussion by sam-eah** [nuxt/nuxt#17243] [Ideas]
I've been really impressed by Nuxt's `useFetch` function and it's ability to return typesafe responses. I think it's a great bonus feature that the other SSR frameworks don't seem to have.

In my opinion, it would be really interesting to give the opportunity to ask for specific body and/or params (query) types in the `defineEventHandler`. This should help not to forget required informations while making queries, and give autocomplete while writing given queries.

Currently, my solution has been to export interfaces like this:

`server/api/user.post.ts`
```ts
export interface CreateUser {
  email: string;
  name: string;
}

export default defineEventHandler(async (event) => {
  const body = await useBody<CreateUser>(event);

  return {
      email: body.email,
      name: body.name,
    };
});
```

`pages/index.vue`
```vue
<script script setup lang="ts">
import { CreateUser } from "../server/api/user.post";

const createUser = async () => {
  const data: CreateUser = { email: "test@test.com", name: "Test" };
  const res = await useFetch("/api/user", {
    method: "post",
    body: data,
  });

  return res
};
</script>
```

I get a strongly typed response, which is really nice. 

![image](https://user-images.githubusercontent.com/15243807/180010674-0d38d7f7-0a02-44e2-9cbe-111a77e28426.png)

However nothing prevents me from giving any body type to the useFetch 😕

Let me know what you think, and if a conceivable improvement is possible!

**Comment by Flanker72** [comment]
Both `useFetch` and `useAsyncData` are generics, so you may define the result type for every request:
```ts
const { data } = await useFetch<IUserData>("/api/user");
```
And this types may be defined as `*.d.ts` within your project to allow using both in the API and in the frontend part.

**Reply by sam-eah** [reply]
This is interesting, but I'm not talking about result types, which are already working out of the box (amazing!) from what I've seen.

My suggestion was more about requests' `body` and `params` types. Currently I haven't found a way to specify required body/params types for the options in the `useFetch`.

I'm sorry if I haven't explained my thoughts clearly 😅

**Reply by arily** [reply]
I ended up using trpc. It’s slow but get the job done.
